src/ui/validate_transaction_nbgl.c

```diff
@@ -0,0 +1,249 @@
+#ifdef HAVE_NBGL
+
+#include "commands.h"
+#include "globals.h"
+#include "glyphs.h"
+#include "menu.h"
+#include "validate_transaction.h"
+#include "nbgl_use_case.h"
+#include "nbgl_page.h"
+#include "nbgl_layout.h"
+#include "os.h"
+#include "io.h"
+#include "ux.h"
+
+#define REFUSAL_TEXT_PART_1 "Incorrect transaction\nrejected by the\n"
+#define REFUSAL_TEXT_PART_2 " app"
+#define REFUSAL_TEXT_MAX_SIZE                                                         \
+    ((sizeof(REFUSAL_TEXT_PART_1) - 1) + (sizeof(G_swap_ctx.payin_binary_name) - 1) + \
+     (sizeof(REFUSAL_TEXT_PART_2) - 1) + 1)
+static char refusal_text[REFUSAL_TEXT_MAX_SIZE];
+
+// One of:
+#define REVIEW_P1_TITLE   "Review transaction"
+#define REVIEW_P1_CONFIRM "Sign transaction"
+
+// Delimitor ' ' or '\n'
+
+#define REVIEW_P2 "to"
+
+// Delimitor ' ' or '\n'
+
+// One of:
+#define REVIEW_P3_SWAP "swap "
+#define REVIEW_P3_SELL "sell "
+#define REVIEW_P3_FUND "fund account"
```

Comment:
no space at the end needed here?

Reply:
No because `dyn_string_1` is empty for FUND, so the next string is the delimitor (` ` or `\n`)
A bit wonky overhaul but it should be done on SDK side, not app side...